src/libtexture/imagecache.cpp

```diff
@@ -3438,18 +3478,154 @@ namespace {
+    result = Strutil::replace(result, "<u>", Strutil::fmt::format("u{}", utile),
+                              true);
+    result = Strutil::replace(result, "<v>", Strutil::fmt::format("v{}", vtile),
+                              true);
+    result = Strutil::replace(result, "<U>",
+                              Strutil::fmt::format("u{}", utile + 1), true);
+    result = Strutil::replace(result, "<V>",
+                              Strutil::fmt::format("v{}", vtile + 1), true);
+    result = Strutil::replace(result, "_u##v##",
+                              Strutil::fmt::format("_u{:02d}v{:02d}", utile,
+                                                   vtile),
+                              true);
+    result = Strutil::replace(result, "%(UDIM)d",
+                              Strutil::fmt::format("{:04d}", udim_tile), true);
+    return result;
+}
+
+
+
+std::string
+ImageCacheFile::udim_to_wildcard(string_view udimpattern)
+{
+    // dot, parens, brackets, and braces are special in regex and need to be
+    // preceded by a backslash to be matched literally.
+    std::string result = Filesystem::filename_to_regex(udimpattern);
+    // Replace the udim specifiers with numeric wildcards, and surround the
+    // number parts with regex parentheses, which will allow us later to
+    // extract just those regions from the match.
+    result = Strutil::replace(result, "<UDIM>", "([0-9]{4})", true);
+    result = Strutil::replace(result, "<u>", "u([0-9]+)", true);
+    result = Strutil::replace(result, "<v>", "v([0-9]+)", true);
+    result = Strutil::replace(result, "<U>", "u([0-9]+)", true);
+    result = Strutil::replace(result, "<V>", "v([0-9]+)", true);
+    result = Strutil::replace(result, "_u##v##", "_u([0-9]+)v([0-9]+)", true);
+    result = Strutil::replace(result, "%(UDIM)d", "([0-9]+)", true);
+    return result;
+}
+
+
+
 ImageCacheFile*
 ImageCacheImpl::resolve_udim(ImageCacheFile* udimfile, Perthread* thread_info,
                              float& s, float& t)
 {
-    // Find the u and v tile IDs, and adjust s,t to take their floors
+    // Find the u and v tile indices
     int utile = std::max(0, int(s));
     int vtile = std::max(0, int(t));
-    s         = s - utile;
-    t         = t - vtile;
 
-    // Synthesized a single combined ID that we'll use as an index.
-    uint64_t id = (uint64_t(vtile) << 32) + uint64_t(utile);
+    // adjust s,t to subtract their floors, leaving only the fractional part
+    s = s - utile;
+    t = t - vtile;
+
+    return resolve_udim(udimfile, thread_info, utile, vtile);
+}
+
+
+
+ImageCacheFile*
+ImageCacheImpl::resolve_udim(ImageCacheFile* udimfile, Perthread* thread_info,
+                             int utile, int vtile)
+{
+    // Synthesize a single combined ID that we'll use as an index into
+    // m_udim_lookup.
+    uint64_t id = udim_id(utile, vtile);
 
     // Which is our mutex from the pool? Use a hash baseed on the filename.
     spin_rw_mutex& udim_lookup_mutex(
```

Comment:
Rather than all this spinlock stuff, couldn't you just resolve all the UDIM's of a file the first time it's requested, in full, store them in a vector (just store null entries for non-existent tiles) and then use constant-time lookup at runtime? The spin_rw_mutex seems expensive to have here.